aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -44,6 +44,9 @@ static inline Tensor& unary_op_impl_out(Tensor& result, const Tensor& self, Stub
   return result;
 }
 
+// out_impl passed into unary_op_impl and unary_op_impl_  must go through at:: device dispatch
+// otherwise it won't dispatch to out-of-source devices like XLA.
+// For example it must be at::bitwise_not_out instead of bitwise_not_out(which is at::native!).
```

Comment:
I'm not sure whether I have understood this. Is it the case that one has to dispatch twice? For example, when `at::bitwise_not` is called, the dispatch should have been done. According to you, it should call `at::bitwise_not_out` instead of `native::bitwise_not_out` to dispatch again.

Reply:
Yea in the current approach it has to dispatch twice since `at::bitwise_not` is a compossible op and `at::bitiwise_not_out` is the op that devices implement.  
To decide an op is compossible, it must have a device dispatch in native_functions.yaml(like bitwise_not_out) or it has a derivative formula defined in derivatives.yaml.

Reply:
I see, does it mean that ops like `sin`, `cos` (which will be migrated here) do not need to call the `at::` version because they have a derivative defined? Is there a recommended practice, such as "calling the `at::` version if possible"? I just wanna make it clear in the doc so we do not make similar mistakes.

Reply:
In general I recommend following the current logic, e.g. for `cos/acos/sin` it actually has `at::` in the generated code. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/UnaryOps.cpp#L238
And in standalone `sign` it also has `at::`. 
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/UnaryOps.cpp#L189
I think there might be some redundancy here, IIUC whether an op is compossible or not is so dynamic (we change native_functions and derivates pretty often) so we always want to dispatch to `at::` to ensure correct dispatch. Please correct me if I'm wrong. @gchanan